# **Data Collection, Cleaning, and Preprocessing for Logistics Analysis**

This notebook documents the data preparation process for logistics analysis.

### **1. Data Collection and Initial Profiling**

This stage establishes the initial structure and quality of the dataset.

**1.1 Import Required Libraries**

The required Python libraries are imported for data analysis.

In [265]:
import pandas as pd
import numpy as np


**1.2 Load the Dataset**

The dataset is loaded into a Pandas DataFrame.

In [389]:
df = pd.read_csv("train.csv")

print("Dataset loaded successfully.")

Dataset loaded successfully.


**1.3 Dataset Dimensions**

The dataset dimensions are examined to determine its size.

In [267]:
rows, columns = df.shape

print("Number of rows:", rows)
print("Number of columns:", columns)

Number of rows: 45593
Number of columns: 20


**1.4 Preview the Dataset**

The first few records are displayed for an initial inspection.

In [268]:
df.head()

,ID,Delivery_person_ID,Delivery_person_Age,Delivery_person_Ratings,Restaurant_latitude,Restaurant_longitude,Delivery_location_latitude,Delivery_location_longitude,Order_Date,Time_Orderd,Time_Order_picked,Weatherconditions,Road_traffic_density,Vehicle_condition,Type_of_order,Type_of_vehicle,multiple_deliveries,Festival,City,Time_taken(min)
0,0x4607,INDORES13DEL02,37,4.9,22.745049,75.892471,22.765049,75.912471,19-03-2022,11:30:00,11:45:00,conditions Sunny,High,2,Snack,motorcycle,0,No,Urban,(min) 24
1,0xb379,BANGRES18DEL02,34,4.5,12.913041,77.683237,13.043041,77.813237,25-03-2022,19:45:00,19:50:00,conditions Stormy,Jam,2,Snack,scooter,1,No,Metropolitian,(min) 33
2,0x5d6d,BANGRES19DEL01,23,4.4,12.914264,77.678400,12.924264,77.688400,19-03-2022,08:30:00,08:45:00,conditions Sandstorms,Low,0,Drinks,motorcycle,1,No,Urban,(min) 26
3,0x7a6a,COIMBRES13DEL02,38,4.7,11.003669,76.976494,11.053669,77.026494,05-04-2022,18:00:00,18:10:00,conditions Sunny,Medium,0,Buffet,motorcycle,1,No,Metropolitian,(min) 21
4,0x70a2,CHENRES12DEL01,32,4.6,12.972793,80.249982,13.012793,80.289982,26-03-2022,13:30:00,13:45:00,conditions Cloudy,High,1,Snack,scooter,1,No,Metropolitian,(min) 30


**1.5 Check Column Names**

The available variables are identified from the dataset.

In [269]:
print(df.columns.tolist())

['ID', 'Delivery_person_ID', 'Delivery_person_Age', 'Delivery_person_Ratings', 'Restaurant_latitude', 'Restaurant_longitude', 'Delivery_location_latitude', 'Delivery_location_longitude', 'Order_Date', 'Time_Orderd', 'Time_Order_picked', 'Weatherconditions', 'Road_traffic_density', 'Vehicle_condition', 'Type_of_order', 'Type_of_vehicle', 'multiple_deliveries', 'Festival', 'City', 'Time_taken(min)']


**1.6 Check Data Types**

The data type of each variable is examined.

In [270]:
df.dtypes

,0
ID,object
Delivery_person_ID,object
Delivery_person_Age,object
Delivery_person_Ratings,object
Restaurant_latitude,float64
Restaurant_longitude,float64
Delivery_location_latitude,float64
Delivery_location_longitude,float64
Order_Date,object
Time_Orderd,object


**1.7 Check Dataset Information**

The dataset structure and non-null counts are examined.

In [271]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45593 entries, 0 to 45592
Data columns (total 20 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   ID                           45593 non-null  object 
 1   Delivery_person_ID           45593 non-null  object 
 2   Delivery_person_Age          45593 non-null  object 
 3   Delivery_person_Ratings      45593 non-null  object 
 4   Restaurant_latitude          45593 non-null  float64
 5   Restaurant_longitude         45593 non-null  float64
 6   Delivery_location_latitude   45593 non-null  float64
 7   Delivery_location_longitude  45593 non-null  float64
 8   Order_Date                   45593 non-null  object 
 9   Time_Orderd                  45593 non-null  object 
 10  Time_Order_picked            45593 non-null  object 
 11  Weatherconditions            45593 non-null  object 
 12  Road_traffic_density         45593 non-null  object 
 13  Vehicle_conditio

**1.8 Generate Statistical Summary**

Descriptive statistics are generated for the dataset.

In [272]:
df.describe(include="all").T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
ID,45593,45593,0x5fb2,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Delivery_person_ID,45593,1320,PUNERES01DEL01,67,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Delivery_person_Age,45593,23,35,2262,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Delivery_person_Ratings,45593,29,4.8,7148,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Restaurant_latitude,45593.0,NaN,NaN,NaN,17.017729,8.185109,-30.905562,12.933284,18.546947,22.728163,30.914057
Restaurant_longitude,45593.0,NaN,NaN,NaN,70.231332,22.883647,-88.366217,73.17,75.898497,78.044095,88.433452
Delivery_location_latitude,45593.0,NaN,NaN,NaN,17.465186,7.335122,0.01,12.988453,18.633934,22.785049,31.054057
Delivery_location_longitude,45593.0,NaN,NaN,NaN,70.845702,21.118812,0.01,73.28,76.002574,78.107044,88.563452
Order_Date,45593,44,15-03-2022,1192,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Time_Orderd,45593,177,NaN,1731,NaN,NaN,NaN,NaN,NaN,NaN,NaN


**1.9 Check Unique Values**

The number of unique values in each variable is examined.

In [273]:
df.nunique()

,0
ID,45593
Delivery_person_ID,1320
Delivery_person_Age,23
Delivery_person_Ratings,29
Restaurant_latitude,657
Restaurant_longitude,518
Delivery_location_latitude,4373
Delivery_location_longitude,4373
Order_Date,44
Time_Orderd,177


### **2. Data Quality Assessment**

The dataset is assessed for common data quality issues before preprocessing.

**2.1 Missing Values**

Missing values are identified across all variables.

In [274]:
missing_values = df.isnull().sum()

print(missing_values)

ID                             0
Delivery_person_ID             0
Delivery_person_Age            0
Delivery_person_Ratings        0
Restaurant_latitude            0
Restaurant_longitude           0
Delivery_location_latitude     0
Delivery_location_longitude    0
Order_Date                     0
Time_Orderd                    0
Time_Order_picked              0
Weatherconditions              0
Road_traffic_density           0
Vehicle_condition              0
Type_of_order                  0
Type_of_vehicle                0
multiple_deliveries            0
Festival                       0
City                           0
Time_taken(min)                0
dtype: int64


**2.2 Calculate Missing Value Percentage**

The percentage of missing values is calculated for each variable.

In [275]:
missing_percentage = (df.isnull().sum() / len(df) * 100).round(2)

missing_summary = pd.DataFrame({
    "Missing Count": df.isnull().sum(),
    "Missing Percentage": missing_percentage
})

missing_summary

,Missing Count,Missing Percentage
ID,0,0.0
Delivery_person_ID,0,0.0
Delivery_person_Age,0,0.0
Delivery_person_Ratings,0,0.0
Restaurant_latitude,0,0.0
Restaurant_longitude,0,0.0
Delivery_location_latitude,0,0.0
Delivery_location_longitude,0,0.0
Order_Date,0,0.0
Time_Orderd,0,0.0


**2.3 Check Hidden Missing Values**

Text fields are checked for common placeholder values representing missing data.

In [276]:
missing_placeholders = ["", " ", "NA", "N/A", "NULL", "null", "NaN", "nan", "None"]

hidden_missing = {}

for column in df.select_dtypes(include="object").columns:
    count = df[column].astype(str).str.strip().isin(missing_placeholders).sum()
    hidden_missing[column] = count

hidden_missing = pd.Series(hidden_missing, name="Hidden Missing Values")

hidden_missing

,Hidden Missing Values
ID,0
Delivery_person_ID,0
Delivery_person_Age,1854
Delivery_person_Ratings,1908
Order_Date,0
Time_Orderd,1731
Time_Order_picked,0
Weatherconditions,0
Road_traffic_density,601
Type_of_order,0


**2.4 Hidden Missing Values**

Placeholder values are identified as hidden missing data.

In [277]:
hidden_missing[hidden_missing > 0]

,Hidden Missing Values
Delivery_person_Age,1854
Delivery_person_Ratings,1908
Time_Orderd,1731
Road_traffic_density,601
multiple_deliveries,993
Festival,228
City,1200


The assessment identified hidden missing values in seven data fields, requiring further treatment during preprocessing.

**2.5 Check Blank and Whitespace Values**

The dataset is checked for blank and whitespace values that may represent missing data.

In [278]:
blank_whitespace = {}

for column in df.select_dtypes(include="object").columns:
    count = df[column].astype(str).str.strip().eq("").sum()
    blank_whitespace[column] = count

blank_whitespace = pd.Series(
    blank_whitespace,
    name="Blank/Whitespace Values"
)

blank_whitespace

,Blank/Whitespace Values
ID,0
Delivery_person_ID,0
Delivery_person_Age,0
Delivery_person_Ratings,0
Order_Date,0
Time_Orderd,0
Time_Order_picked,0
Weatherconditions,0
Road_traffic_density,0
Type_of_order,0


**2.6 Validate Record Identifiers**

Record identifiers are checked for unexpected duplication.

In [279]:
print("Duplicate IDs:", df["ID"].duplicated().sum())
print("Unique IDs:", df["ID"].nunique())

Duplicate IDs: 0
Unique IDs: 45593


**2.7 Suspicious Numerical Values**

Numerical ranges are examined to identify potentially invalid values.

In [280]:
numeric_columns = df.select_dtypes(include=np.number).columns

for column in numeric_columns:
    print(f"\n{column}")
    print("Minimum:", df[column].min())
    print("Maximum:", df[column].max())


Restaurant_latitude
Minimum: -30.905562
Maximum: 30.914057

Restaurant_longitude
Minimum: -88.366217
Maximum: 88.433452

Delivery_location_latitude
Minimum: 0.01
Maximum: 31.054057

Delivery_location_longitude
Minimum: 0.01
Maximum: 88.563452

Vehicle_condition
Minimum: 0
Maximum: 3


**2.8 Data Quality Assessment Summary**

The identified issues are recorded for appropriate preprocessing.

In [281]:
quality_summary = pd.DataFrame({
    "Column": df.columns,
    "Data Type": df.dtypes.astype(str),
    "Missing Values": df.isnull().sum(),
    "Unique Values": df.nunique()
})

quality_summary

,Column,Data Type,Missing Values,Unique Values
ID,ID,object,0,45593
Delivery_person_ID,Delivery_person_ID,object,0,1320
Delivery_person_Age,Delivery_person_Age,object,0,23
Delivery_person_Ratings,Delivery_person_Ratings,object,0,29
Restaurant_latitude,Restaurant_latitude,float64,0,657
Restaurant_longitude,Restaurant_longitude,float64,0,518
Delivery_location_latitude,Delivery_location_latitude,float64,0,4373
Delivery_location_longitude,Delivery_location_longitude,float64,0,4373
Order_Date,Order_Date,object,0,44
Time_Orderd,Time_Orderd,object,0,177


## **3. Missing-Value Handling**

**3.1 Standardize Missing Values**

Hidden missing-value placeholders are converted into standard NaN values for consistent processing.

In [361]:
# Define common missing-value placeholders

missing_placeholders = [
    "", " ", "NA", "N/A", "NULL",
    "null", "NaN", "nan", "None"
]

# Convert hidden missing values to NaN

for column in df.select_dtypes(include="object").columns:
    df[column] = df[column].replace(missing_placeholders, np.nan)

    df[column] = df[column].apply(
        lambda x: np.nan
        if isinstance(x, str) and x.strip() == ""
        else x
    )

print("Hidden missing values standardized successfully.")

Hidden missing values standardized successfully.


**3.2 Recheck Missing Values**

The dataset is rechecked to identify actual missing values after standardization.

In [362]:
# Check hidden missing values after standardization

missing_values = df.isnull().sum()

print("Missing values identified after standardization:")
print(missing_values[missing_values > 0])

Missing values identified after standardization:
Series([], dtype: int64)


No standard missing values were identified at this stage. Additional missing values may be generated during numeric and datetime conversion and will be handled in the subsequent cleaning steps.


**3.3 Calculate Missing-Value Percentage**

The percentage of missing values is calculated to assess their extent across the dataset.

In [363]:
# Calculate current missing-value summary

missing_summary = pd.DataFrame({
    "Missing Count": df.isnull().sum(),
    "Missing Percentage": (
        df.isnull().sum() / len(df) * 100
    ).round(2)
})

print("Initial missing-value summary:")
display(missing_summary[missing_summary["Missing Count"] > 0])

Initial missing-value summary:


,Missing Count,Missing Percentage


**3.4 Identify Columns Requiring Treatment**

Columns containing missing observations are identified for appropriate treatment.

In [420]:
# Identify columns currently containing missing values

missing_columns = missing_summary[
    missing_summary["Missing Count"] > 0
].index.tolist()

print("Columns currently containing missing values:")
print(missing_columns)


Columns currently containing missing values:
[]


No missing values are currently present. Additional missing values generated during data-type conversion will be identified and treated in the subsequent cleaning steps.

**3.5 Missing-Value Treatment Strategy**


Numerical variables will be treated using median imputation after data-type conversion. Categorical variables will be treated using mode imputation. Missing order times will be handled using the mode after datetime conversion.

In [365]:
print("Missing-value treatment strategy defined.")
print("Numerical variables: Median")
print("Categorical variables: Mode")
print("Order time: Mode")

Missing-value treatment strategy defined.
Numerical variables: Median
Categorical variables: Mode
Order time: Mode


**3.6 Treatment Timing**


Missing-value treatment is intentionally postponed until data-type conversion. This ensures that values becoming missing during numeric or datetime conversion are also included in the final treatment process.

In [366]:
print("Missing-value treatment will be completed after data-type conversion.")

Missing-value treatment will be completed after data-type conversion.


**3.7 Pre-Conversion Verification**

The dataset is checked before data-type conversion. Final verification will be performed after conversion and missing-value treatment.

In [367]:
# Verify missing values before data-type conversion

pre_conversion_missing = df.isnull().sum()

print("Missing values before data-type conversion:")
print(pre_conversion_missing[pre_conversion_missing > 0])

print("\nFinal missing-value verification will be performed after conversion.")

Missing values before data-type conversion:
Series([], dtype: int64)

Final missing-value verification will be performed after conversion.


**3.8 Initial Missing-Value Summary**


An initial missing-value summary is recorded before data-type conversion. Additional missing values identified during conversion will be handled in the subsequent cleaning steps.

In [368]:
# Initial missing-value summary

initial_missing_summary = pd.DataFrame({
    "Missing Count": df.isnull().sum(),
    "Missing Percentage": (
        df.isnull().sum() / len(df) * 100
    ).round(2)
})

initial_missing_summary

,Missing Count,Missing Percentage
ID,0,0.0
Delivery_person_ID,0,0.0
Delivery_person_Age,0,0.0
Delivery_person_Ratings,0,0.0
Restaurant_latitude,0,0.0
Restaurant_longitude,0,0.0
Delivery_location_latitude,0,0.0
Delivery_location_longitude,0,0.0
Order_Date,0,0.0
Time_Orderd,0,0.0


### **4. Duplicate and Data Consistency Handling**

**4.1 Check Duplicate Records**

Duplicate records are checked to identify repeated delivery transactions that could affect analysis and model results.

In [369]:
# Check for duplicate records

duplicate_count = df.duplicated().sum()

print("Number of duplicate records:", duplicate_count)

Number of duplicate records: 0


**4.2 Identify Duplicate Records**

Even though the count is zero, keep a separate verification step.

Duplicate rows are identified for further inspection before removal

In [370]:
# Identify duplicate rows

duplicate_rows = df[df.duplicated(keep=False)]

print("Duplicate rows identified:", len(duplicate_rows))

duplicate_rows.head()

Duplicate rows identified: 0


,ID,Delivery_person_ID,Delivery_person_Age,Delivery_person_Ratings,Restaurant_latitude,Restaurant_longitude,Delivery_location_latitude,Delivery_location_longitude,Order_Date,Time_Orderd,Time_Order_picked,Weatherconditions,Road_traffic_density,Vehicle_condition,Type_of_order,Type_of_vehicle,multiple_deliveries,Festival,City,Time_taken(min)


**4.3 Remove Duplicate Records**


Duplicate records are removed to ensure that each delivery transaction is represented only once.

In [371]:
# Remove duplicate records

before_duplicates = len(df)

df = df.drop_duplicates().reset_index(drop=True)

after_duplicates = len(df)

print("Rows before duplicate removal:", before_duplicates)
print("Rows after duplicate removal:", after_duplicates)
print("Duplicate rows remaining:", df.duplicated().sum())

Rows before duplicate removal: 45593
Rows after duplicate removal: 45593
Duplicate rows remaining: 0


**4.4 Validate Record Count**


The dataset dimensions are verified after duplicate handling to confirm that the expected records have been retained.

In [372]:
# Validate dataset dimensions after duplicate handling

print("Dataset shape after duplicate handling:", df.shape)
print("Total records retained:", len(df))

Dataset shape after duplicate handling: (45593, 20)
Total records retained: 45593


**4.5 Duplicate Handling Summary**


No duplicate records were found in the dataset. Therefore, all 45,593 delivery records were retained for subsequent data cleaning and preprocessing.

In [373]:
# Duplicate handling summary

duplicate_summary = pd.DataFrame({
    "Metric": [
        "Rows Before Duplicate Handling",
        "Duplicate Rows Identified",
        "Rows After Duplicate Handling",
        "Duplicate Rows Remaining"
    ],
    "Result": [
        before_duplicates,
        duplicate_count,
        after_duplicates,
        df.duplicated().sum()
    ]
})

duplicate_summary

,Metric,Result
0,Rows Before Duplicate Handling,45593
1,Duplicate Rows Identified,0
2,Rows After Duplicate Handling,45593
3,Duplicate Rows Remaining,0


### **5. Data-Type Conversion and Cleaning**

**5.1 Convert Numeric Columns**


Numeric fields stored as text are converted to numeric data types for accurate calculations and analysis.

In [391]:
# Convert numeric columns

numeric_columns = [
    "Delivery_person_Age",
    "Delivery_person_Ratings",
    "multiple_deliveries"
]

for column in numeric_columns:
    df[column] = pd.to_numeric(df[column], errors="coerce")

# Clean and convert Time_taken(min)
df["Time_taken(min)"] = (
    df["Time_taken(min)"]
    .astype(str)
    .str.extract(r"(\d+)", expand=False)
)

df["Time_taken(min)"] = pd.to_numeric(
    df["Time_taken(min)"],
    errors="coerce"
)

print("Numeric columns converted successfully.")

Numeric columns converted successfully.


**5.2 Convert Order Date**

The order date is converted to a datetime format to support date-based logistics analysis.

In [392]:
# Convert order date to datetime

df["Order_Date"] = pd.to_datetime(
    df["Order_Date"],
    dayfirst=True,
    errors="coerce"
)

print("Order date converted successfully.")

Order date converted successfully.


**5.3 Convert Order Times**


Order and pickup times are converted into time-compatible formats for delivery-time analysis.

In [393]:
# Convert order and pickup times to datetime format

df["Time_Orderd"] = pd.to_datetime(
    df["Time_Orderd"],
    format="%H:%M:%S",
    errors="coerce"
)

df["Time_Order_picked"] = pd.to_datetime(
    df["Time_Order_picked"],
    format="%H:%M:%S",
    errors="coerce"
)

print("Order and pickup times converted successfully.")

Order and pickup times converted successfully.


**5.4 Validate Data-Type Conversion**


Converted columns are validated to confirm that the required data types were applied successfully.

In [394]:
# Validate converted data types

converted_columns = [
    "Delivery_person_Age",
    "Delivery_person_Ratings",
    "multiple_deliveries",
    "Time_taken(min)",
    "Order_Date",
    "Time_Orderd",
    "Time_Order_picked"
]

print(df[converted_columns].dtypes)

Delivery_person_Age               float64
Delivery_person_Ratings           float64
multiple_deliveries               float64
Time_taken(min)                     int64
Order_Date                 datetime64[ns]
Time_Orderd                datetime64[ns]
Time_Order_picked          datetime64[ns]
dtype: object


 **5.5 Inspect Converted Order Times**

The converted order and pickup times are inspected to confirm that the time values have been correctly interpreted after datetime conversion.


In [390]:
print(df[["Time_Orderd", "Time_Order_picked"]].head(10))

  Time_Orderd Time_Order_picked
0    11:30:00          11:45:00
1    19:45:00          19:50:00
2    08:30:00          08:45:00
3    18:00:00          18:10:00
4    13:30:00          13:45:00
5    21:20:00          21:30:00
6    19:15:00          19:30:00
7    17:25:00          17:30:00
8    20:55:00          21:05:00
9    21:55:00          22:10:00


**5.6 Treat Invalid Ratings**

Delivery ratings are validated against the expected 1–5 scale. Ratings above 5 are treated as invalid and converted to NaN so that they can be handled consistently with other missing values.

In [396]:
# Identify invalid delivery ratings

invalid_ratings = (
    df["Delivery_person_Ratings"] > 5
).sum()

print("Invalid ratings above 5:", invalid_ratings)

# Convert invalid ratings to NaN

df.loc[
    df["Delivery_person_Ratings"] > 5,
    "Delivery_person_Ratings"
] = np.nan

print(
    "Missing ratings after invalid-value treatment:",
    df["Delivery_person_Ratings"].isnull().sum()
)

Invalid ratings above 5: 53
Missing ratings after invalid-value treatment: 1961


**5.7 Handle Missing Values After Data-Type Conversion**


Missing numerical values are treated using median imputation to reduce the influence of extreme observations. Missing categorical values are treated using mode imputation. Missing order times are handled using the most frequent valid time.

In [397]:
# Handle missing numerical values using median imputation

numerical_missing_columns = [
    "Delivery_person_Age",
    "Delivery_person_Ratings",
    "multiple_deliveries"
]

for column in numerical_missing_columns:
    if df[column].isnull().sum() > 0:
        df[column] = df[column].fillna(
            df[column].median()
        )


# Handle missing categorical values using mode imputation

categorical_missing_columns = [
    "Road_traffic_density",
    "Festival",
    "City"
]

for column in categorical_missing_columns:
    if df[column].isnull().sum() > 0:
        df[column] = df[column].fillna(
            df[column].mode()[0]
        )


# Handle missing order times using the most frequent valid time

if df["Time_Orderd"].isnull().sum() > 0:
    df["Time_Orderd"] = df["Time_Orderd"].fillna(
        df["Time_Orderd"].mode()[0]
    )

print("Missing-value treatment completed.")

Missing-value treatment completed.


In [398]:
remaining_missing = df.isnull().sum()

print("Remaining missing values:")
print(remaining_missing[remaining_missing > 0])

Remaining missing values:
Series([], dtype: int64)


### **6. Outlier Detection and Treatment**

**6.1 Identify Numerical Columns**


The Interquartile Range (IQR) method is used to identify potential outliers in key numerical variables. These observations are reviewed before treatment because extreme values may represent genuine logistics conditions.

In [399]:
# Identify potential outliers using the IQR method

outlier_columns = [
    "Delivery_person_Age",
    "Delivery_person_Ratings",
    "multiple_deliveries",
    "Time_taken(min)"
]

outlier_summary = []

for column in outlier_columns:
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outlier_count = (
        (df[column] < lower_bound) |
        (df[column] > upper_bound)
    ).sum()

    outlier_summary.append({
        "Column": column,
        "Q1": Q1,
        "Q3": Q3,
        "Lower Bound": lower_bound,
        "Upper Bound": upper_bound,
        "Potential Outliers": outlier_count
    })

outlier_summary_df = pd.DataFrame(outlier_summary)

outlier_summary_df

,Column,Q1,Q3,Lower Bound,Upper Bound,Potential Outliers
0,Delivery_person_Age,25.0,34.0,11.5,47.5,53
1,Delivery_person_Ratings,4.6,4.8,4.3,5.1,5292
2,multiple_deliveries,0.0,1.0,-1.5,2.5,361
3,Time_taken(min),19.0,32.0,-0.5,51.5,270


**6.2 Review Potential Outliers**

IQR screening identified potential extreme observations. These values are reviewed against the expected business range before treatment. Values that are operationally valid are retained rather than removed automatically.

In [400]:
# Validate key variables against their expected business ranges

range_validation = pd.DataFrame({
    "Variable": [
        "Delivery_person_Age",
        "Delivery_person_Ratings",
        "multiple_deliveries",
        "Time_taken(min)"
    ],
    "Minimum": [
        df["Delivery_person_Age"].min(),
        df["Delivery_person_Ratings"].min(),
        df["multiple_deliveries"].min(),
        df["Time_taken(min)"].min()
    ],
    "Maximum": [
        df["Delivery_person_Age"].max(),
        df["Delivery_person_Ratings"].max(),
        df["multiple_deliveries"].max(),
        df["Time_taken(min)"].max()
    ]
})

range_validation

,Variable,Minimum,Maximum
0,Delivery_person_Age,15.0,50.0
1,Delivery_person_Ratings,1.0,5.0
2,multiple_deliveries,0.0,3.0
3,Time_taken(min),10.0,54.0


**6.3 — Outlier Treatment Decision**


Based on range validation, the identified extreme observations fall within plausible operational ranges. Therefore, no records are removed solely based on IQR screening. Retaining these observations preserves genuine variations in delivery operations and avoids losing useful information for subsequent logistics analysis.

In [401]:
# Confirm that key numerical variables contain no invalid range values

invalid_range_counts = {
    "Delivery_person_Age": (
        (df["Delivery_person_Age"] < 15) |
        (df["Delivery_person_Age"] > 50)
    ).sum(),

    "Delivery_person_Ratings": (
        (df["Delivery_person_Ratings"] < 1) |
        (df["Delivery_person_Ratings"] > 5)
    ).sum(),

    "multiple_deliveries": (
        (df["multiple_deliveries"] < 0) |
        (df["multiple_deliveries"] > 3)
    ).sum(),

    "Time_taken(min)": (
        (df["Time_taken(min)"] < 10) |
        (df["Time_taken(min)"] > 54)
    ).sum()
}

invalid_range_counts

{'Delivery_person_Age': np.int64(0),
 'Delivery_person_Ratings': np.int64(0),
 'multiple_deliveries': np.int64(0),
 'Time_taken(min)': np.int64(0)}

**6.4 Validate Outlier Treatment**

The outlier treatment is validated to confirm that delivery ratings now fall within the expected 1–5 range and that no missing ratings remain after imputation. Legitimate extreme values in other numerical variables are retained.


In [416]:
# Validate rating treatment

print("Ratings above 5:", (df["Delivery_person_Ratings"] > 5).sum())
print("Ratings below 1:", (df["Delivery_person_Ratings"] < 1).sum())
print("Missing ratings:", df["Delivery_person_Ratings"].isnull().sum())

Ratings above 5: 0
Ratings below 1: 0
Missing ratings: 0


### **7. Data Normalization**

**7.1 Apply Min-Max Normalization**


Min-Max normalization scales selected numerical variables to a common range of 0 to 1. A separate normalized dataset is created while preserving the original cleaned dataset.

In [403]:
from sklearn.preprocessing import MinMaxScaler

normalization_columns = [
    "Delivery_person_Age",
    "Delivery_person_Ratings",
    "multiple_deliveries",
    "Time_taken(min)"
]

scaler = MinMaxScaler()

df_normalized = df.copy()

df_normalized[normalization_columns] = scaler.fit_transform(
    df[normalization_columns]
)

print("Min-Max normalization applied successfully.")

Min-Max normalization applied successfully.


**7.2 Validate Normalized Values**


The normalized numerical variables are validated to confirm that their values have been successfully scaled to the expected 0 to 1 range.

In [417]:
normalization_validation = pd.DataFrame({
    "Column": normalization_columns,
    "Minimum": [
        df_normalized[column].min()
        for column in normalization_columns
    ],
    "Maximum": [
        df_normalized[column].max()
        for column in normalization_columns
    ]
})

normalization_validation

,Column,Minimum,Maximum
0,Delivery_person_Age,0.0,1.0
1,Delivery_person_Ratings,0.0,1.0
2,multiple_deliveries,0.0,1.0
3,Time_taken(min),0.0,1.0


**7.3 Compare Original and Normalized Values**


Original and normalized values are compared to confirm that normalization preserves the relative structure of the data.

In [405]:
# Compare original and normalized values

comparison = pd.DataFrame({
    "Original Age": df["Delivery_person_Age"].head(5),
    "Normalized Age": df_normalized["Delivery_person_Age"].head(5),
    "Original Rating": df["Delivery_person_Ratings"].head(5),
    "Normalized Rating": df_normalized["Delivery_person_Ratings"].head(5),
    "Original Time": df["Time_taken(min)"].head(5),
    "Normalized Time": df_normalized["Time_taken(min)"].head(5)
})

comparison

,Original Age,Normalized Age,Original Rating,Normalized Rating,Original Time,Normalized Time
0,37.0,0.628571,4.9,0.975,24,0.318182
1,34.0,0.542857,4.5,0.875,33,0.522727
2,23.0,0.228571,4.4,0.850,26,0.363636
3,38.0,0.657143,4.7,0.925,21,0.250000
4,32.0,0.485714,4.6,0.900,30,0.454545


### **8. Preprocessing Validation**

**8.1 Validate Dataset Structure**


The preprocessed dataset is checked to confirm that the original number of records and variables has been maintained after cleaning and transformation.

In [406]:
# Validate final dataset structure

print("Rows:", df.shape[0])
print("Columns:", df.shape[1])
print("Duplicate rows:", df.duplicated().sum())

Rows: 45593
Columns: 20
Duplicate rows: 0


**8.2 Missing-Value Validation**


The cleaned dataset is checked to confirm that all identified missing values have been successfully handled during preprocessing.

In [408]:
# Check remaining missing values

remaining_missing = df.isnull().sum()

print("Remaining missing values:")
print(remaining_missing[remaining_missing > 0])

Remaining missing values:
Series([], dtype: int64)


**8.3 Missing-Value Treatment Validation**


The preprocessing results are validated to confirm that no unexpected missing values remain in the final dataset.

In [409]:
# Validate that no missing values remain

total_missing = df.isnull().sum().sum()

print("Total remaining missing values:", total_missing)

if total_missing == 0:
    print("Missing-value validation passed.")
else:
    print("Missing-value validation failed.")

Total remaining missing values: 0
Missing-value validation passed.


**8.4 Data-Type Validation**


The final data types are checked to confirm that numerical, date, and time fields remain in the expected formats after preprocessing.

In [411]:
# Validate final data types

validation_columns = [
    "Delivery_person_Age",
    "Delivery_person_Ratings",
    "multiple_deliveries",
    "Time_taken(min)",
    "Order_Date",
    "Time_Orderd",
    "Time_Order_picked"
]

print(df[validation_columns].dtypes)

Delivery_person_Age               float64
Delivery_person_Ratings           float64
multiple_deliveries               float64
Time_taken(min)                     int64
Order_Date                 datetime64[ns]
Time_Orderd                datetime64[ns]
Time_Order_picked          datetime64[ns]
dtype: object


**8.5 Final Preprocessing Integrity Check**

A final integrity check confirms that the cleaned dataset retains all 45,593 delivery records, contains 20 variables, has no duplicate records or missing delivery-time values, and is ready for further logistics analysis.

In [418]:
# Final preprocessing integrity check

print("Final rows:", df.shape[0])
print("Final columns:", df.shape[1])
print("Duplicate rows:", df.duplicated().sum())
print("Remaining missing values:", df.isnull().sum().sum())
print(
    "Missing delivery time values:",
    df["Time_taken(min)"].isnull().sum()
)

Final rows: 45593
Final columns: 20
Duplicate rows: 0
Remaining missing values: 0
Missing delivery time values: 0


### **9. Final Data Quality Summary**

**9.1 Create Final Quality Summary**


The final data quality summary consolidates the key preprocessing results, including dataset size, duplicate records, missing values, invalid ratings, and delivery-time completeness.

In [419]:
# Create final data quality summary

final_quality_summary = pd.DataFrame({
    "Metric": [
        "Total Rows",
        "Total Columns",
        "Duplicate Rows",
        "Remaining Missing Values",
        "Invalid Ratings Identified and Treated",
        "Invalid Ratings Remaining",
        "Missing Delivery Time Values"
    ],
    "Result": [
        df.shape[0],
        df.shape[1],
        df.duplicated().sum(),
        df.isnull().sum().sum(),
        invalid_ratings,
        (df["Delivery_person_Ratings"] > 5).sum(),
        df["Time_taken(min)"].isnull().sum()
    ]
})

final_quality_summary

,Metric,Result
0,Total Rows,45593
1,Total Columns,20
2,Duplicate Rows,0
3,Remaining Missing Values,0
4,Invalid Ratings Identified and Treated,53
5,Invalid Ratings Remaining,0
6,Missing Delivery Time Values,0


**9.2 Final Dataset Preview**


A final preview is used to confirm that the cleaned dataset is ready for downstream logistics analysis and predictive modeling.

In [414]:
# Preview the cleaned dataset

df.head()

,ID,Delivery_person_ID,Delivery_person_Age,Delivery_person_Ratings,Restaurant_latitude,Restaurant_longitude,Delivery_location_latitude,Delivery_location_longitude,Order_Date,Time_Orderd,Time_Order_picked,Weatherconditions,Road_traffic_density,Vehicle_condition,Type_of_order,Type_of_vehicle,multiple_deliveries,Festival,City,Time_taken(min)
0,0x4607,INDORES13DEL02,37.0,4.9,22.745049,75.892471,22.765049,75.912471,2022-03-19,1900-01-01 11:30:00,1900-01-01 11:45:00,conditions Sunny,High,2,Snack,motorcycle,0.0,No,Urban,24
1,0xb379,BANGRES18DEL02,34.0,4.5,12.913041,77.683237,13.043041,77.813237,2022-03-25,1900-01-01 19:45:00,1900-01-01 19:50:00,conditions Stormy,Jam,2,Snack,scooter,1.0,No,Metropolitian,33
2,0x5d6d,BANGRES19DEL01,23.0,4.4,12.914264,77.678400,12.924264,77.688400,2022-03-19,1900-01-01 08:30:00,1900-01-01 08:45:00,conditions Sandstorms,Low,0,Drinks,motorcycle,1.0,No,Urban,26
3,0x7a6a,COIMBRES13DEL02,38.0,4.7,11.003669,76.976494,11.053669,77.026494,2022-04-05,1900-01-01 18:00:00,1900-01-01 18:10:00,conditions Sunny,Medium,0,Buffet,motorcycle,1.0,No,Metropolitian,21
4,0x70a2,CHENRES12DEL01,32.0,4.6,12.972793,80.249982,13.012793,80.289982,2022-03-26,1900-01-01 13:30:00,1900-01-01 13:45:00,conditions Cloudy,High,1,Snack,scooter,1.0,No,Metropolitian,30


### **10. Reflection and Logistics Impact**



The Week 2 preprocessing stage transformed the original Food Delivery dataset into a structured and analysis-ready dataset containing **45,593 delivery records and 20 variables**. No duplicate records remained after validation, and all identified missing values were handled during preprocessing.

A total of **53 invalid delivery-rating values above the expected 1–5 range were identified and treated**. Potential outliers were reviewed using IQR analysis and operational range validation. Legitimate extreme observations were retained to avoid removing genuine delivery variations.

Numerical variables were converted to appropriate formats, date and time fields were standardized, and selected numerical variables were normalized using Min-Max scaling for future analytical and predictive modeling.

These preprocessing steps improve the reliability of subsequent logistics analysis by providing a consistent dataset for studying **delivery time, traffic conditions, weather, vehicle conditions, driver-related factors, and other operational variables**. The cleaned dataset can now be used for exploratory analysis and predictive modeling in the following stages.
